semgrep-core/toy-matcher/Matcher.ml

```diff
@@ -0,0 +1,562 @@
+                             function.
+     See the documentation for the Hashtbl module for more info.
+  *)
+  let hash (ellipsis, env, pat, input) =
+    Hashtbl.hash_param 10 100 ellipsis
+    + Hashtbl.hash_param 10 100 pat
+    + Hashtbl.hash_param 10 100 input
+    + hash_env env
+end
+
+module Memoize = struct
+  module Tbl = Hashtbl.Make (Cache_key)
+
+  (* only use cache once in 3 times *)
+  let cache_every = 3
+
+  let should_use_cache input =
+    Hashtbl.hash_param 5 10 input mod cache_every = 0
+
+  let get tbl compute ellipsis env pat input =
+    (* only use the cache on some inputs because it's expensive *)
+    if should_use_cache input then
+      let key = (ellipsis, env, pat, input) in
+      match Tbl.find_opt tbl key with
+      | None ->
+          let res = compute ellipsis env pat input in
+          Tbl.add tbl key res;
+          res
+      | Some res -> res
+    else
+      compute ellipsis env pat input
+
+  let create compute =
+    (*
+       Initial table size impact performance.
+    *)
+    let tbl = Tbl.create 8192 in
+    fun ellipsis env pat input ->
+      get tbl compute ellipsis env pat input
+end
+
+(*
+   Main matching function.
+
+   Checks if a pattern matches the entire input sequence.
+   Returns the captured symbols or sequences of symbols for which a
+   name was specified in the pattern.
+*)
+let match_input ?(trace = true) ?(cache = false) root_pat root_input =
+  let stat = { match_calls = 0 } in
+  let get_from_cache = ref (fun _ellipsis _env _pat _input -> assert false) in
+
+  let rec match_ (ellipsis : ellipsis) (env : env) pat input =
+    trace_match_call ~trace stat ellipsis env pat input;
+    !get_from_cache ellipsis env pat input
+
+  and uncached_match ellipsis env pat input =
+    let orig_pat = pat in
+    let orig_input = input in
+    let in_ellipsis = ellipsis <> None in
+    match pat with
+    | [] ->
+        (match input with
+         | [] ->
+             let env = close_ellipsis ellipsis env in
+             Some env
+         | symbol :: input ->
+             if in_ellipsis then
+               let ellipsis = extend_ellipsis ellipsis symbol in
+               match_ ellipsis env orig_pat input
+             else
+               None
+        )
+    | (pat_atom, opt_name) :: pat ->
+        match input with
+        | [] ->
+            (* end of input, only empty sequence can match *)
+            (match pat_atom with
+             | Ellipsis ->
+                 let env = extend env ellipsis opt_name [] in
+                 match_ (init_ellipsis opt_name []) env pat input
+             | Backref name ->
+                 (match Env.find_opt name env with
+                  | Some [] ->
+                      let env = extend env ellipsis opt_name [] in
+                      match_ None env pat input
+                  | _ ->
+                      None
+                 )
+             | _ -> None
```

Comment:
Maybe better to be explicit here, given there's not so many cases in pattern_atom:
| Any_symbol  | Symbol _ -> None

Reply:
done